parlai/core/worlds.py

```diff
@@ -1245,6 +1252,127 @@ def shutdown(self):
+        self.produce_queue = produce_queue
+        self.consume_queue = consume_queue
+        self.produce_cond = produce_cond
+        self.consume_cond = consume_cond
+
+    def parley(self):
+        """
+        Produce a batch observation.
+
+        Consume a batch act.
+        """
+        if hasattr(self.world, 'parley_init'):
+            for w in self.worlds:
+                w.parley_init()
+        teacher_idx = 0
+        agent_idx = 1
+        with self.produce_cond:
+            self.produce_cond.wait_for(self.produce_queue.empty)
+            batch_act = self.batch_act(teacher_idx, None)
+            batch_obs = self.batch_observe(agent_idx, batch_act, teacher_idx)
+            self.produce_queue.put(batch_obs)
+        with self.consume_cond:
+            self.consume_cond.wait_for(self.consume_queue.full)
+            batch_act = self.consume_queue.get()
+            for other_idx in range(len(self.world.get_agents())):
+                self.batch_observe(other_idx, batch_act, agent_idx)
+        self.update_counters()
+
+
+def run_q_world(opt, world, mp_structs):
+    world = QBatchWorld(
+        opt,
+        world,
+        mp_structs['produce_queue'],
+        mp_structs['consume_queue'],
+        mp_structs['produce_cond'],
+        mp_structs['consume_cond'],
+    )
+    while not world.epoch_done():
+        world.parley()
+
+
+class QueueWorld(BatchWorld):
+    """
+    QueueWorld creates subprocess worlds.
+
+    They create their own batches.
+    """
+
+    def __init__(self, opt: Opt, world: World):
+        World.__init__(self, opt)
+        # QueueWorld init
+        self.world: World = world
+        self.processes: List[Process] = []
+        self.structs: List[Dict[str, Union[Queue, Condition]]] = []
+        self.worlds: List[World] = []
+        for _ in range(opt['numworkers']):
+            produce_cond = Condition()
+            consume_cond = Condition()
+            produce_queue = Queue(maxsize=1)
+            consume_queue = Queue(maxsize=1)
+            shared = world.share()
+            subworld = shared['world_class'](opt, None, shared)
+            self.worlds.append(subworld)
+            self.structs.append(
+                {
+                    'produce_queue': produce_queue,
+                    'consume_queue': consume_queue,
+                    'produce_cond': produce_cond,
+                    'consume_cond': consume_cond,
+                }
+            )
+            self.processes.append(
+                Process(
+                    target=run_q_world,
+                    args=(opt, subworld, self.structs[-1]),
+                    daemon=True,
+                )
+            )
+        for p in self.processes:
+            p.start()
+
+    def parley(self):
+        batch_obs: List[Message] = None
+        chosen_p: Dict[str, Union[Queue, Condition]] = None
+        # Get batch obs
+        while batch_obs is None:
+            chosen_p = random.choice(self.structs)
```

Comment:
I think this is why we aren't getting the speedup. We should have a SINGLE produce q, which all workers can put onto. This world should be reading *solely* off that queue.